### ISSUE TITLE
[1871] Server and client differ on player's cash

### ISSUE THREAD

**Issue by ljphillips84** [tobymao/18xx#8004]
**What's your Game ID?**

https://18xx.games/game/87284

- Chat shows where it happened (e.g. https://18xx.games/game/87284?action=305 and the turn or so before) - Souris was split to create Stratford Branch (SB), I tried to buy a share but I get an error band flash at the top of the screen:
![Screenshot 2022-06-24 8 05 33 PM](https://user-images.githubusercontent.com/10617839/175649412-2f4d19ce-80c5-4839-b343-05931491fbc9.png)

It then resets the game to before the buy action.

I can't see any rule or reason why I should be prevented from buying, and it doesn't happen when I clone the game.

**How to Reproduce the Bug?**

Trying to buy a share of SB, a recently formed company from splitting. I could exchange Ice boat shipping for a share so owning one isn't prevented, just purchasing one. I'm not near the cert limit. Other players can buy shares no problem.

**Which Browser / Platform?**
Tried on multiple, same on all:
- Edge / PC
- Chrome / Chrome OS tablet
- Firefox / android

**Comment by dfannius** [comment]
The typical underlying cause of this sort of error (player is trying to buy something and has barely enough cash but the server rejects it) is that a floating-point discrepancy between the server and the client means that the server thinks the player has less money than the client thinks.

TOP 1871 did have a clear problem of this sort. I thought I fixed it with https://github.com/tobymao/18xx/pull/7983 but there must still be an error lurking.

I do notice this in the log:
```
[10:49] PEIR runs for $50 and pays half
[10:49] PEIR withholds $24
[10:49] PEIR pays out $26 = $7 per share ($14 to Iyidin_Kyeimo [UTC+0], $14 to serpentium)
```
which is suspicious. It probably wasn't the problem all by itself (by the rules, PEIR should have withheld $25 but still paid $7 for each share because $25/4 rounded up is $7) but points to there still being a floating-point issue somewhere.

**Comment by dfannius** [comment]
(Half-paying was incorrect and is fixed with #8005, but it was just a logic error that probably executed the same on both the server and the client, so I don't think it's at fault here.)

**Comment by dfannius** [comment]
This may be fixed by #8011. The code there definitely was not working as intended.

**Comment by dfannius** [comment]
Wasn't really fixed, see #8040. (But it's about to be).
